### PR TITLE
APEXCORE-474 In M*1 case, deploy the unifier in the same container as…

### PR DESCRIPTION
**Pull request by sandeshh**
… downstream

@tweise and @PramodSSImmaneni Please review.

### CODE REVIEW

File: engine/src/test/java/com/datatorrent/stram/StreamingContainerManagerTest.java [hunk 1]
```diff
@@ -322,10 +322,21 @@ public void testStaticPartitioning()
     }
 
     // unifier
-    List<PTOperator> o2Unifiers = plan.getMergeOperators(dag.getMeta(node2));
-    Assert.assertEquals("number unifiers", 1, o2Unifiers.size());
-    List<OperatorDeployInfo> cUnifier = getDeployInfo(dnm.getContainerAgent(o2Unifiers.get(0).getContainer().getExternalId()));
-    Assert.assertEquals("number operators " + cUnifier, 1, cUnifier.size());
+    PTContainer containerWithUnifier;
```
Comment (tweise) [inline engine/src/test/java/com/datatorrent/stram/StreamingContainerManagerTest.java]: How about getting the container of the downstream operator instead?

**Review by tweise — CHANGES_REQUESTED** [review]

File: engine/src/test/java/com/datatorrent/stram/StreamingContainerManagerTest.java [hunk 2]
```diff
@@ -764,15 +775,20 @@ public void testOperatorShutdown()
     PTOperator o3p1 = physicalPlan.getOperators(dag.getMeta(o3)).get(0);
     MockContainer mc4 = mockContainers.get(o3p1.getContainer());
     MockOperatorStats o3p1mos = mc4.stats(o3p1.getId());
-    o3p1mos.currentWindowId(1).checkpointWindowId(1).deployState(DeployState.ACTIVE);
-    mc4.sendHeartbeat();
+    PTOperator unifier = null;
```
Comment (tweise) [inline engine/src/test/java/com/datatorrent/stram/StreamingContainerManagerTest.java]:
```
PTOperator unifier = o3p1.upstreamMerge.values().iterator().next();
```

**Review by tweise — CHANGES_REQUESTED** [review]

File: engine/src/test/java/com/datatorrent/stram/plan/physical/PhysicalPlanTest.java [hunk 3]
```diff
@@ -668,7 +671,8 @@ public void testRepartitioningScaleDownSinglePartition()
     Assert.assertSame("", p1Doper.getOperatorMeta(), o1Meta.getMeta(o1.output).getUnifierMeta());
     Assert.assertTrue("unifier ", p1Doper.isUnifier());
 
-    Collection<PTOperator> o1Unifiers = plan.getMergeOperators(o1Meta);
+    Collection<PTOperator> o1Unifiers = getUnifierStartingWithName(plan, o1Meta.getName());
```
Comment (tweise) [inline engine/src/test/java/com/datatorrent/stram/plan/physical/PhysicalPlanTest.java]: No point having this here. You could add assert above for

```
o1p1.getOutputs().get(0).sinks.size()
```

**Review by tweise — CHANGES_REQUESTED** [review]

File: engine/src/test/java/com/datatorrent/stram/StreamingContainerManagerTest.java (hunk 1 above)
Comment (sandeshh) [reply engine/src/test/java/com/datatorrent/stram/StreamingContainerManagerTest.java]: Done.

File: engine/src/test/java/com/datatorrent/stram/StreamingContainerManagerTest.java (hunk 2 above)
Comment (sandeshh) [reply engine/src/test/java/com/datatorrent/stram/StreamingContainerManagerTest.java]: Done.

File: engine/src/test/java/com/datatorrent/stram/plan/physical/PhysicalPlanTest.java (hunk 3 above)
Comment (sandeshh) [reply engine/src/test/java/com/datatorrent/stram/plan/physical/PhysicalPlanTest.java]: Done.

File: engine/src/main/java/com/datatorrent/stram/plan/physical/StreamMapping.java [hunk 4]
```diff
@@ -307,6 +307,11 @@ private void redoMapping()
           } else {
             // MxN partitioning: unifier per downstream partition
             LOG.debug("MxN unifier for {} {} {}", new Object[] {doperEntry.first, doperEntry.second.getPortName(), pks});
+
+            if (pks != null && pks.mask == 0) {
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/plan/physical/StreamMapping.java]: Maybe move that down so it matches previous pattern?

**Review by tweise — CHANGES_REQUESTED** [review]
I'm going to send you a PR with more changes.

**Pull request merged**